cmd/snap-bootstrap/cmd_initramfs_mounts.go

```diff
@@ -91,6 +103,126 @@ func stampedAction(stamp string, action func() error) error {
+	// is the default behavior
+	NoWait bool
+}
+
+var (
+	doSystemdMount = doSystemdMountImpl
+	counter        = 1
+)
+
+// doSystemdMount will mount "what" at "where" using systemd-mount(1) with
+// various options
+func doSystemdMountImpl(what, where string, opts *SystemdMountOptions) error {
+	if opts == nil {
+		opts = &SystemdMountOptions{}
+	}
+
+	// doesn't make sense to fsck a tmpfs
+	if opts.NeedsFsck && opts.IsTmpfs {
+		return fmt.Errorf("cannot mount %q at %q: impossible to fsck a tmpfs", what, where)
+	}
+
+	whereEscaped := systemd.EscapeUnitNamePath(where)
+	unitName := whereEscaped + ".mount"
+
+	args := []string{"--no-pager", "--no-ask-password", what, where}
+	if opts.IsTmpfs {
+		args = append(args, "--type=tmpfs")
+	}
+
+	if opts.NeedsFsck {
+		args = append(args, "--fsck=yes")
+	}
+
+	// TODO: do we really need this? unclear why we would need to ever not block
+	if opts.NoWait {
+		args = append(args, "--no-block")
+	}
+
+	// note that we do not currently parse any output from systemd-mount, but if
+	// we ever do, take special care surrounding the debug output that systemd
+	// outputs with the "debug" kernel command line present (or equivalently the
+	// SYSTEMD_LOG_LEVEL=debug env var) which will add lots of additional output
+	// to stderr from systemd commands
+	out, err := exec.Command("systemd-mount", args...).CombinedOutput()
+	if err != nil {
+		return osutil.OutputErr(out, err)
+	}
+
+	// finally if it should survive pivot_root() then we need to add symlinks
+	// to /run/systemd
+	if opts.SurvivesPivotRoot {
+		// to survive the pivot_root, we need to make the mount units depend on
+		// all of the various initrd special targets by adding runtime conf
+		// files there
+		// note we could do this statically in the initramfs main filesystem
+		// layout, but that means that changes to snap-bootstrap would block on
+		// waiting for those files to be added before things works here, this is
+		// a more flexible strategy that puts snap-bootstrap in control
+		for _, initrdUnit := range []string{
+			"initrd.target",
+			"initrd-fs.target",
+			"initrd-switch-root.target",
+		} {
+			targetDir := filepath.Join("/run/systemd/system", initrdUnit+".d")
+			err := os.MkdirAll(targetDir, 0755)
+			if err != nil {
+				return err
+			}
+
+			// add an override file for the initrd unit to depend on this mount
+			// unit so that when we isolate to the initrd unit, it does not get
+			// unmounted
+			fname := fmt.Sprintf("snap_bootstrap_%s.conf", whereEscaped)
+			content := []byte(fmt.Sprintf(unitFileDependOverride, unitName))
+			err = ioutil.WriteFile(filepath.Join(targetDir, fname), content, 0644)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	if !opts.NoWait {
+		// TODO: is this necessary, systemd-mount seems to only return when the
+		// unit is active and the mount is there, but perhaps we should be a bit
+		// paranoid here and wait anyways?
+		// see systemd-mount(1)
```

Comment:
But also fsck, no? We should waif for the fsck to start & finish.

Reply:
ah good point, I hadn't considered that will be simply started, but may not be finished yet. I will add some code waiting for the fsck to finish after we finish waiting for the mount to show up (still being extra paranoid).

Reply:
Actually, from systemd-mount (emphasis mine):

>  --fsck=
>       Takes a boolean argument, defaults to on. **Controls whether to run a file system check
>       immediately before the mount operation.** In the automount case (see --automount= below)
>       the check will be run the moment the first access to the device is made, which might
>       slightly delay the access.

So systemd will not perform/start the mount until the fsck unit has finished and returned successfully. So it's already blocking on the fsck operation before doing the mount, so I don't think we need to do anything else here. 

This can also be confirmed manually by running systemd-mount with `--fsck=yes` and then looking at the generated mount unit,

```
$ sudo systemd-mount /dev/mapper/loop85p3 /tmp/p3 --fsck=yes
Started unit tmp-p3.mount for mount point: /tmp/p3
$ systemctl cat tmp-p3.mount 
# /run/systemd/transient/tmp-p3.mount
# This is a transient unit file, created programmatically via the systemd API. Do not edit.
[Unit]
After=dev-dm\x2d2.device
BindsTo=dev-dm\x2d2.device

[Mount]
What=/dev/dm-2

[Unit]
Requires=systemd-fsck@dev-dm\x2d2.service
After=systemd-fsck@dev-dm\x2d2.service
```

Reply:
Well, fsck=yes is the default. I'm more concerned to ensure that we wait for it to finish and try to repair the filesystem when fsck fails.

Reply:
so if the fsck unit fails, what should we do here, as snap-bootstrap? right now with the current status of the this PR, snap-bootstrap will exit with non-zero exit code causing the-tool to just loop infinitely waiting for output, but when we fix the-tool to not interpret the output of `snap-bootstrap initramfs-mounts`, the-tool will die and we will just sit in the initramfs, possibly opening a debug shell if `dangerous` is defined. That matches what happens on master without this PR as well IIUC.

IMHO, I don't think we should block this PR on taking restorative action if the fsck fails as this PR doesn't make that situation worse and it will be easier to handle as a followup when we have a concrete set of actions to take if fsck fails.